### ISSUE TITLE
Need an easy way to add logging

### ISSUE THREAD

**Issue by dalewking** [orbit-mvi/orbit-mvi#157]
I'm coming from MVIKotlin and like the simpler nature of Orbit. But one thing MVIKotlin did well was made it easy to add logging to the MVI store. It probably would be pretty easy to do logging of state changes and side effects using a context decorator but what would be missing is logging of intents.

I know logging was mentioned in the medium article as a future plan but don't see anything in the code to support it yet

**Comment by dalewking** [comment]
Have not had time to look into it in detail other than to see that it looks like you are using a compiler plugin. Can you summarize the strategy here?

**Comment by mattmook** [comment]
In terms of logging, we want to trigger the logger every time we call `reduce` or `sideEffect`. But for that to be useful you need to know what `intent` block was called and potentially any data used in the triggering. You of course want to log the relevant state or sideEffect data.

Additionally you could have multiple orbit containers running at the same time so it is useful to know the ContainerHost name too.

What this means is we roughly want the following `Logger` interface (realise I missed checking the Logger class into the branch above):

```
public interface Logger {
    public fun logState(containerHostName: String, intentName: String, state: Any)
    public fun logSideEffect(containerHostName: String, intentName: String, sideEffect: Any)
}
```

We can relatively easily expose a name parameter in the `intent` block, called `intentName`, for the user of the library to provide a name. However, it doesn't scale well. This is where a compiler plugin can help by not only looking up the function name the intent block is within but also inserting any data into that string. For example, imagine a calculator where the user can press on a digit, we might have an intent as follows:

```
fun digitPressed(digit: Int) = intent { ... }
```

The compiler plugin will convert that into:

```
fun digitPressed(digit: Int) = intent(intentName = "digitPressed(digit = $digit)") { ... }
```

You can of course override that name manually if you so wish... basically if `intentName` is already provided in the call to `intent` then the plugin does nothing.

Similarly when you create the `ContainerHost` there's a new `containerHostName` argument that is automatically populated by the compiler plugin based on the name of the class that defines the `ContainerHost`.

**Comment by mattmook** [comment]
Coming back to your original message... it would be relatively easy for us to add in logIntent to the Logger too:

```
public fun logIntent(containerHostName: String, intentName: String)
```

**Comment by dalewking** [comment]
> In terms of logging, we want to trigger the logger every time we call reduce or sideEffect. But for that to be useful you need to know what intent block was called and potentially any data used in the triggering. You of course want to log the relevant state or sideEffect data.

Logging reduce and sideEffect are pretty trivial right now. That can be done with a ContainerDecorator that passes a ContainerContext that does the logging before calling the real container. You wouldn't get the intent, but I'm not sure how vital that is, especially if you log the intent.

The intent is the hard case. It is much easier in other APIs like MVI Kotlin where intents are objects and they go through the same data flow.

Compiler plugin idea is interesting, but almost seems killing a mosquito with a bazooka.

I wonder if something could be done with declaring intents as properties with a functional type and property delegation is used to get the name. So you declare intents more like this:

`val digitPressed: (Int) -> Unit by Intent { digit -> ... }`

Not sure if that is workable as it is off the top of my head.

**Comment by Rosomack** [comment]
Interesting idea!

While I like it, the reason we went for a compiler plugin is that it's completely transparent to the developer. Adding new APIs is not frictionless, especially as the community is already used to a particular way of working with the library.

**Comment by mattmook** [comment]
This also moves away from one of the unofficial goals of Orbit, to be free of boilerplate. While delegates might work there's a fair bit of boilerplate everytime you want to use one.

While writing a compiler plugin wasn't high on my list of things to do,. I felt it helped fit the logging (and eventually time-travel debugging) use case quite nicely. The names used can be customised if the user so wishes but for most, a simple inclusion in your project and you get the nice names for free with no visible boilerplate. Granted the plugin will no doubt work in all our users use cases, but if it caters for the 80% then we've done pretty well.

I'm curious, using a delegate while it is easy to retrieve the properties name you would need to use reflection to query any parameters provided?

**Comment by dalewking** [comment]
So I finally took the time to dig into the PR and it basically is only giving me one thing more than I can already do using a ContainerDecorator. As I said it is rather trivial to make a ContainerDecorator that adds logging to reduce and postSideEffect calls.

But what is missing there is logging of intents. The compiler plugin is giving logging of intents and getting name and parameter values automatically, but it puts them in the log statements for state changes and side effects. That produces lots of text for both. It also logs nothing if you have an intent that ends up not doing anything which still may be something I want to know about.

I would rather see the intent logged separately and not made part of state and side effect logging.

Also may want to think about logging other things like container creation, onCreate execution, or subscription

Note that it computes that text for intent always with no optimization for turning logging off, so that is a performance concern. We would want logging on for debug and not production so we don't want to waste time creating the string if it isn't used.

**Comment by mattmook** [comment]
As I was mentioning before given your points we should be adding a `logIntent` call to the logger in addition to what is already there.

To me, logging `reduce` and `postSideEffect` execution without knowing the context (i.e. the intent) isn't always that helpful.

For a lot of use cases the `intent` logged independently with `reduce`/`postSideEffect` not including that information may be okay. But, when you have a slightly more complex example where multiple `intents` trigger simultaneously can you always easily tell what `intent` triggered what `reduce`? This is a multi-threaded system after all. That's why having the `intent` with the `reduce`/`postSideEffect` feels important.

You are right about computation and production mode. Not something I'd given much thought to which is exactly the purpose of creating an early draft PR and getting feedback.

It's definitely not great always creating a string especially when, as you say, it may not even be used. I guess we could implement it as a lazy message similar to how Kotlin's `require` deals with its message:

```
public inline fun require(value: Boolean, lazyMessage: () -> Any): Unit {
    if (!value) {
        val message = lazyMessage()
        throw IllegalArgumentException(message.toString())
    }
}
```

Of course, arguably, we may not want to leak those strings into production code either. I do wonder if there's a clever trick with ProGuard/R8 that could be used here? `-assumevalues` perhaps (https://r8.googlesource.com/r8/+/refs/tags/1.5.45/src/test/proguard/valid/assume-values-with-return-value.flags / https://jakewharton.com/r8-optimization-value-assumption/)

**Comment by dalewking** [comment]
> To me, logging reduce and postSideEffect execution without knowing the context (i.e. the intent) isn't always that helpful.

Correct me if I'm wrong but other than inlineOrbit calls aren't intents executed sequentially so that no intents are executing simultaneously? If so then the context is the preceding intent call above it in the log as long as you don't log the intent before it starts executing. If not then you have an unworkable system that will lead to race conditions and data integrity problems.

I am currently using MVIKotlin and that is how it works.

The only exception is the inline intent but I think it would be a very bad idea to ever call reduce from an inline intent unless there is something I am missing where they don't execute concurrently. The only thing I would use an inline intent for are those cases where an intent does nothing but emit a side effect.

Perhaps on the performance/leaking question perhaps we could only enable the compiler plugin in debug mode. Proguard is not an answer since we have more than just android

**Comment by mattmook** [comment]
No not exactly. They are executed in order up to a suspension point in the code. (Under the hood we use an [unconfined dispatcher](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-unconfined.html))

Intents blocking can lead to long running network requests blocking your UI from responding until they complete. In effect the `reduce` calls block which gets around most integrity problems, and state.copy(...) is your friend here too. Of course you have to be careful with how you update the data (state.copy changing different parts of the state works well for example) in the cases the you know you'll have multiple calls happening at the same time.

We use Orbit in some pretty large code bases and haven't had any issues with that regard.

**Comment by TWiStErRob** [comment]
I'm just starting to use Orbit, I came up with this, and it seems to do everything @dalewking is asking for:
```kotlin
abstract class BaseViewModel<State : Any, Effect : Any>(
	initialState: State
) : ViewModel(), ContainerHost<State, Effect> {

	override val container = container<State, Effect>(initialState)
		// One line installation project-wide
		// (if there's a base class, or otherwise centralized `container` creation)
		.decorateLogging(LoggerFactory.getLogger(this::class.java))
}
```

<details><summary>Click to see reflective decorator implementation</summary>

```kotlin
import org.orbitmvi.orbit.Container
import org.orbitmvi.orbit.ContainerDecorator
import org.orbitmvi.orbit.annotation.OrbitInternal
import org.orbitmvi.orbit.syntax.ContainerContext
import org.orbitmvi.orbit.syntax.simple.SimpleContext
import org.orbitmvi.orbit.syntax.simple.SimpleSyntax

@OptIn(OrbitInternal::class)
class LoggingContainerDecorator<STATE : Any, SIDE_EFFECT : Any>(
	override val actual: Container<STATE, SIDE_EFFECT>,
	private val events: OrbitEvents<STATE, SIDE_EFFECT>,
) : ContainerDecorator<STATE, SIDE_EFFECT> {

	interface OrbitEvents<STATE : Any, SIDE_EFFECT : Any> {
		fun intent(transformer: suspend SimpleSyntax<STATE, SIDE_EFFECT>.() -> Unit)

		fun sideEffect(sideEffect: SIDE_EFFECT)

		fun reduce(oldState: STATE, reducer: SimpleContext<STATE>.() -> STATE, newState: STATE)
	}

	override suspend fun orbit(orbitIntent: suspend ContainerContext<STATE, SIDE_EFFECT>.() -> Unit) {
		super.orbit {
			events.intent(orbitIntent.captured("transformer"))
			this.logged().orbitIntent()
		}
	}

	override suspend fun inlineOrbit(orbitIntent: suspend ContainerContext<STATE, SIDE_EFFECT>.() -> Unit) {
		super.inlineOrbit {
			events.intent(orbitIntent.captured("transformer"))
			this.logged().orbitIntent()
		}
	}

	/**
	 * @see org.orbitmvi.orbit.syntax.simple.intent
	 * @see org.orbitmvi.orbit.syntax.simple.blockingIntent
	 * @see org.orbitmvi.orbit.syntax.simple.reduce
	 */
	private fun ContainerContext<STATE, SIDE_EFFECT>.logged(): ContainerContext<STATE, SIDE_EFFECT> =
		ContainerContext(
			settings = settings,
			postSideEffect = {
				events.sideEffect(it)
				postSideEffect(it)
			},
			getState = ::state,
			reduce = { reducer ->
				reduce { oldState ->
					reducer(oldState).also { newState ->
						events.reduce(oldState, reducer.captured("reducer"), newState)
					}
				}
			},
			subscribedCounter = subscribedCounter,
		)
}

/**
 * The framework's reducer and intent declarations are lambdas inside Orbit's functions.
 * We need to access the captured local variables to get our original reducer and intent lambdas.
 */
private fun <T : Function<*>> Function<*>.captured(localName: String): T =
	this::class
		.java
		.getDeclaredField("\$${localName}")
		.apply { isAccessible = true }
		.get(this)
		.let { @Suppress("UNCHECKED_CAST") (it as T) }
```

```kotlin
import net.twisterrob.orbit.logging.LoggingContainerDecorator.OrbitEvents
import org.orbitmvi.orbit.Container
import org.orbitmvi.orbit.syntax.simple.SimpleContext
import org.orbitmvi.orbit.syntax.simple.SimpleSyntax
import org.slf4j.Logger

class OrbitSlf4jLogger<STATE : Any, SIDE_EFFECT : Any>(
	private val log: Logger
) : OrbitEvents<STATE, SIDE_EFFECT> {

	override fun intent(
		transformer: suspend SimpleSyntax<STATE, SIDE_EFFECT>.() -> Unit
	) {
		log.trace("Starting intent: {} with {}", transformer.lambdaName(), transformer.captures())
	}

	override fun reduce(
		oldState: STATE,
		reducer: SimpleContext<STATE>.() -> STATE,
		newState: STATE
	) {
		log.trace("reduced via {}:\n\t{}\n\t->\n\t{}", reducer.lambdaName(), oldState, newState)
	}

	override fun sideEffect(
		sideEffect: SIDE_EFFECT
	) {
		log.trace("postSideEffect({})", sideEffect)
	}

	companion object {
		fun <S : Any, SE : Any> Container<S, SE>.decorateLogging(log: Logger): Container<S, SE> =
			LoggingContainerDecorator(this, OrbitSlf4jLogger(log))
	}
}

/**
 * An instantiated (not inlined) lambda is always an instance of Function,
 * it'll have the enclosing class AND method name as the class name.
 */
private fun Function<*>.lambdaName(): String = this::class.java.name
	// Lambdas (anonymous classes) don't have .simpleName,
	// so need to resort to string manipulation to remove the package prefix.
	.replaceBeforeLast('.', "")
	// Remove the last dot too.
	.removePrefix(".")

/**
 * When a lambda is called in Kotlin, it creates an anonymous inner class.
 * This anonymous inner class has fields for the captured variables.
 * When launching the intent, the only parameters captured should be the inputs of the event.
 */
private fun Function<*>.captures(): Map<String, Any?> =
	this::class
		.java
		.declaredFields
		.filter { it.name.startsWith("$") }
		.onEach { it.isAccessible = true }
		.associateBy({ it.name.removePrefix("$") }, { it.get(this) })
```

</details>

Example output:
<details><summary>Click to see MVI setup</summary>

```kotlin
@HiltViewModel
internal class ManageSpaceViewModel @Inject constructor(
	private val useCase: GetSizesUseCase,
	private val mapper: SizesDomainToStateMapper,
) : BaseViewModel<ManageSpaceState, ManageSpaceEffect>(
	initialState = ManageSpaceState(
		isLoading = false,
		sizes = null,
		dialog = null,
	)
) {
	fun loadSizes() { // Called on page load, pull to refresh and from rebuildSearch below.
		intent {
			reduce {
				state.copy(
					isLoading = true,
					sizes = mapper.loading(),
				)
			}
			val input = useCase.execute(Unit)
			val sizes = mapper.map(input)
			reduce {
				state.copy(
					isLoading = false,
					sizes = sizes,
				)
			}
		}
	}
	fun rebuildSearchConfirmed() { // Called from user interaction.
		intent {
			reduce {
				state.copy(
					dialog = DialogState("Rebuild search index…"),
				)
			}
		}
	}

	fun rebuildSearch(activity: Activity) { // Called when dialog is dismissed with "OK".
		intent {
			reduce {
				state.copy(
					dialog = null,
				)
			}
			Database.get(activity.applicationContext).rebuildSearch()
			loadSizes()
		}
	}
}
```
</details>

```
Starting intent: ManageSpaceViewModel$loadSizes$1 with {}

reduced via ManageSpaceViewModel$loadSizes$1$1:
	ManageSpaceState(isLoading=false, sizes=null, dialog=null)
	->
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)

reduced via ManageSpaceViewModel$loadSizes$1$2:
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)

Starting intent: ManageSpaceViewModel$rebuildSearchConfirmed$1 with {}

reduced via ManageSpaceViewModel$rebuildSearchConfirmed$1$1:
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=DialogState(title=Rebuild search index…))

Starting intent: ManageSpaceViewModel$rebuildSearch$1 with {activity=net.twisterrob.inventory.android.activity.space.ManageSpaceActivity@1eae988}

reduced via ManageSpaceViewModel$rebuildSearch$1$1:
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=DialogState(title=Rebuild search index…))
	->
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)

Starting intent: ManageSpaceViewModel$loadSizes$1 with {}

reduced via ManageSpaceViewModel$loadSizes$1$1:
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)

reduced via ManageSpaceViewModel$loadSizes$1$2:
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)

Starting intent: ManageSpaceViewModel$loadSizes$1 with {}

reduced via ManageSpaceViewModel$loadSizes$1$1:
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)

reduced via ManageSpaceViewModel$loadSizes$1$2:
	ManageSpaceState(isLoading=true, sizes=SizesState(imageCache=Calculating…, database=Calculating…, freelist=Calculating…, searchIndex=Calculating…, allData=Calculating…, errors=null), dialog=null)
	->
	ManageSpaceState(isLoading=false, sizes=SizesState(imageCache=157 KB, database=292 KB, freelist=0.00 B, searchIndex=1.09 KB, allData=1.09 MB, errors=null), dialog=null)
```

**Comment by dalewking** [comment]
> I'm just starting to use Orbit, I came up with this, and it seems to do everything @dalewking is asking for:

Haven't dug into your code completely, but there are certainly ways to do it. As I said before reduce and side effects are easy

Your intent logging leaves a bit to be desired. We have only done a POC of switching to Orbit from MVIKotlin. What I was thinking of doing is based around the way I was planning on doing my intents. I like the idea of a sealed hierarchy of data classes and objects to represent Intents like we have in MVIKotlin. That makes a nice toString for the intent from one object.

So I would define something like this for your example:
```
    sealed class Intention {
        object: LoadSizes : Intention()
        object RebuildSearchConfirmed: Intention()
	data class RebuildSearch(val activity: Activity): Intention()
    }
```

My intention methods would then all be overloads of the invoke operator function:

```
	operator fun invoke(intention: LoadSizes) { ... }
	operator fun invoke(intention: RebuildSearchConfirmed) { ... }
	operator fun invoke(intention: RebuildSearch) { ... }
```

Then I was going to use extensions or sub interfaces to define the concept of passing the intention to intent methods for logging purposes, but I have not gotten around to trying to code it yet as it is still just a proof of concept for our project.

**Comment by TWiStErRob** [comment]
Ah, I see. I think the above might work in that case too, there'll be always a way to reach the `intention` object, because otherwise the body of the `invoke` calling `intent` wouldn't be able to access it either. This workaround heavily relies on Kotlin internals/reflection and is meant for the Orbit-recommended way of declaring intents, which doesn't have sealed classes, but it should work for your case too with some potential modifications. (https://www.youtube.com/watch?v=q7y6NQ_ou0A describes it quite well, why we don't really need sealed classes to model "function calls" - that's why I chose this lib -, but I digress.) I agree that the `$...$1` is not pretty, but I hope it might help someone adapt my code to their liking.

I would love to have some easy-to-use built-in feature though, as this reflective hack is quite the code to maintain in any real project. The compiler plugin idea sounds good, I'm curious if it works out.
